src/proj.rs

```diff
@@ -0,0 +1,791 @@
+    ];
+
+    Some(normalize(inv))
+}
+
+fn mul3x3(a: [f32; 9], b: [f32; 9]) -> [f32; 9] {
+    let (
+        a11, a12, a13,
+        a21, a22, a23,
+        a31, a32, a33
+        ) = (
+        a[0], a[1], a[2],
+        a[3], a[4], a[5], 
+        a[6], a[7], a[8]
+        );
+    let (
+        b11, b12, b13,
+        b21, b22, b23,
+        b31, b32, b33
+        ) = (
+        b[0], b[1], b[2],
+        b[3], b[4], b[5], 
+        b[6], b[7], b[8]
+        );
+
+    [a11*b11 + a12*b21 + a13*b31, a11*b12 + a12*b22 + a13*b32, a11*b13 + a12*b23 +a13*b33,
+     a21*b11 + a22*b21 + a23*b31, a21*b12 + a22*b22 + a23*b32, a21*b13 + a22*b23 + a23*b33,
+     a31*b11 + a32*b21 + a33*b31, a31*b12 + a32*b22 + a33*b32, a31*b13 + a32*b23 + a33*b33]
+}
+
+fn blend<P>(
+    top_left: P,
+    top_right: P,
+    bottom_left: P,
+    bottom_right: P,
+    right_weight: f32,
+    bottom_weight: f32,
+) -> P
+where
+    P: Pixel,
+    P::Subpixel: ValueInto<f32> + Clamp<f32>,
+{
+    let top = top_left.map2(&top_right, |u, v| {
+        P::Subpixel::clamp((1f32 - right_weight) * cast(u) + right_weight * cast(v))
+    });
+
+    let bottom = bottom_left.map2(&bottom_right, |u, v| {
+        P::Subpixel::clamp((1f32 - right_weight) * cast(u) + right_weight * cast(v))
+    });
+
+    top.map2(&bottom, |u, v| {
+        P::Subpixel::clamp((1f32 - bottom_weight) * cast(u) + bottom_weight * cast(v))
+    })
+}
+
+
+#[allow(deprecated)]
+fn interpolate<P>(image: &Image<P>, x: f32, y: f32, default: P) -> P
+where
+    P: Pixel + 'static,
+    <P as Pixel>::Subpixel: ValueInto<f32> + Clamp<f32>,
+{
+    let left = x.floor();
+    let right = left + 1f32;
+    let top = y.floor();
+    let bottom = top + 1f32;
+
+    let right_weight = x - left;
+    let bottom_weight = y - top;
+
+    // default if out of bound
+    let (width, height) = image.dimensions();
+    if left < 0f32 || right >= width as f32 || top < 0f32 || bottom >= height as f32 {
+        default
+    } else {
+        let (tl, tr, bl, br) = unsafe {
+            (
+                image.unsafe_get_pixel(left as u32, top as u32),
+                image.unsafe_get_pixel(right as u32, top as u32),
+                image.unsafe_get_pixel(left as u32, bottom as u32),
+                image.unsafe_get_pixel(right as u32, bottom as u32),
+            )
+        };
+        blend(tl, tr, bl, br, right_weight, bottom_weight)
+    }
+}
+
+#[inline(always)]
+#[allow(deprecated)]
+fn nearest<P: Pixel + 'static>(image: &Image<P>, x: f32, y: f32, default: P) -> P {
+    let rx = x.round() as u32;
+    let ry = y.round() as u32;
+    let (width, height) = image.dimensions();
+    if x.round() < 0f32 ||  y.round() < 0f32 {
+        default
+    } else if rx >= width || ry >= height {
+        default
+    } else {
+        unsafe { image.unsafe_get_pixel(rx, ry) }
+    }
+}
+
+/// How to handle pixels whose pre-image lies between input pixels.
+#[derive(Debug, PartialEq, Eq, Copy, Clone)]
+pub enum Interpolation {
```

Comment:
Use the type from `affine`.

Maybe combine affine.rs and proj.rs into a new geometric_transformation.rs? We should keep the specialised functions for translation, rotation, etc., as we can make them faster than a general projective transformation, but I'm not sure that we then need a special case for Affine transformations - those can be subsumed into your new projection functions.

Reply:
I like this suggestion. Thought about that before, but didn't want to seem too pushy, hence the separate module. I will combine the two in geometric_trasnformation.rs.

That will make a breaking change in the crate, so I guess after that a version bump is really needed.

Reply:
Thanks. Yes, I’ll publish a new version after this PR is landed.

Reply:
@theotherphil, finished merging affine and proj module. Took affine::translate verbatim, as it was super fast. Reimplemented rotate by using Projection + warp.
Removed most of the helper functions that play around HasBlack trait.

I wanted to use try_invert with f64 - for better numeric stability, but it was crashing regression tests (affine_bilinear_rgb) so I gave up on that.

- warp* with projection take in->out homography as was the case with affine.
- reverted order of parameters so in image is first, and homography is 2nd (to be more like affine)